### PR TITLE
Add gunicorn timeout

### PR DESCRIPTION
**Pull request by matherit**
*Issue #, if available:*

*Description of changes:*
Makes the GUnicorn startup timeout configurable via an environment variable. I also updated the readme to include a section that details configurable env vars like this one that do not fit under current sections.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by josephevans — APPROVED** [review]
LGTM, thanks for updating the docs too!

File: docker/build_artifacts/sagemaker/serve.py [hunk 1]
```diff
@@ -449,7 +452,7 @@ def start(self):
             self._setup_gunicorn()
             self._start_gunicorn()
             # make sure gunicorn is up
-            with self._timeout(seconds=30):
+            with self._timeout(seconds=self._gunicorn_setup_timeout_seconds):
```
Comment (satishpasumarthi) [inline docker/build_artifacts/sagemaker/serve.py:455]: You would need to update the `gunicorn_command` as well to consume the timeout. This comment refers to line # 202 in this file where gunicorn command is setup.

File: README.md [hunk 2]
```diff
@@ -612,6 +614,21 @@ SAGEMAKER_TFS_NUM_BATCH_THREADS="16"
...
+```bash
+# Configures the logging level for GUnicorn.
+# Valid values can be found here: https://docs.gunicorn.org/en/stable/settings.html#loglevel
+# Defaults to "info"
+SAGEMAKER_GUNICORN_LOGLEVEL="debug"
+
+# Configures how long to wait in seconds for GUnicorn
+# to finish starting up before timing out.
+# Defaults to 30.
+SAGEMAKER_GUNICORN_SETUP_TIMEOUT_SECONDS="60"
```
Comment (satishpasumarthi) [inline README.md:629]: Add a link to the actual option of gunicorn. https://docs.gunicorn.org/en/stable/settings.html#timeout
Example: You can mention something like `for setting this parameter, refer: https://docs.gunicorn.org/en/stable/settings.html#timeout`

File: README.md [hunk 3]
```diff
@@ -612,6 +614,21 @@ SAGEMAKER_TFS_NUM_BATCH_THREADS="16"
...
 ```
 
+## Other Configurable Environment Variables
+The following environment variables can be set on a SageMaker Model or Transform Job if further configuration is required:
+
+```bash
+# Configures the logging level for GUnicorn.
+# Valid values can be found here: https://docs.gunicorn.org/en/stable/settings.html#loglevel
+# Defaults to "info"
+SAGEMAKER_GUNICORN_LOGLEVEL="debug"
```
Comment (satishpasumarthi) [inline README.md:624]: Add reference link here https://docs.gunicorn.org/en/stable/settings.html#loglevel
See my below comment for more details.

File: README.md [hunk 4]
```diff
@@ -42,7 +42,9 @@ For notebook examples, see: [Amazon SageMaker Examples](https://github.com/awsla
 3. [Running the tests](#running-the-tests)
 4. [Pre/Post-Processing](#pre/post-processing)
 5. [Deploying a TensorFlow Serving Model](#deploying-a-tensorflow-serving-model)
-6. [Deploying to Multi-Model Endpoint](#deploying-to-multi-model-endpoint)
+6. [Enable Batching](#enabling-batching)
+7. [Other Configurable Environment Variables](#other-configurable-environment-variables)
```
Comment (satishpasumarthi) [inline README.md:46]: Maybe rename to `Configurable SageMaker Environment Variables` ? Since these are SageMaker job specific?

**Review by satishpasumarthi — CHANGES_REQUESTED** [review]
Thanks for the PR. I've left some comments. Please take a look

File: docker/build_artifacts/sagemaker/serve.py (hunk 1 above)
Comment (matherit) [reply docker/build_artifacts/sagemaker/serve.py:461]: Good catch. As a result, this PR now includes two env vars. One for the `gunicorn_command` and one for this timeout. This timeout may end up being unnecessary, but it is probably better than just a hardcoded value.

Comment (satishpasumarthi) [reply docker/build_artifacts/sagemaker/serve.py:461]: Both are effectively the same. The code is using `SIGALARM` to kill the process if the `gunicorn`  process is not up in the specified time (previously hard coded `30` seconds). You should replace `self._gunicorn_setup_timeout_seconds` with `self._gunicorn_timeout_seconds`

File: README.md [hunk 5]
```diff
@@ -612,6 +614,23 @@ SAGEMAKER_TFS_NUM_BATCH_THREADS="16"
...
+```bash
+# Configures the logging level for GUnicorn.
+# When looking to set this environment variable, please refer to:
+# https://docs.gunicorn.org/en/stable/settings.html#loglevel
+# Defaults to "info"
+SAGEMAKER_GUNICORN_LOGLEVEL="debug"
+
+# Configures how long a GUnicorn worker may be silent before it is killed and restarted.
+# When looking to set this environment variable, please refer to:
+# https://docs.gunicorn.org/en/stable/settings.html#timeout
```
Comment (satishpasumarthi) [inline README.md:629]: Please see if you can make it a clickable link.

File: README.md [hunk 6]
```diff
@@ -612,6 +614,23 @@ SAGEMAKER_TFS_NUM_BATCH_THREADS="16"
 SAGEMAKER_TFS_MAX_ENQUEUED_BATCHES="10000"
 ```
 
+## Configurable SageMaker Environment Variables
+The following environment variables can be set on a SageMaker Model or Transform Job if further configuration is required:
+
+```bash
+# Configures the logging level for GUnicorn.
+# When looking to set this environment variable, please refer to:
+# https://docs.gunicorn.org/en/stable/settings.html#loglevel
```
Comment (satishpasumarthi) [inline README.md:623]: Please see if you can make it a clickable link.

Comment (matherit) [reply README.md]: Let me know how you feel about my current readme. It removes the `please refer to` language, but I felt that would just be unnecessary "boilerplate" language.

I'm open to swapping it back if we feel that having that was actually better though. 

Thanks!

Comment (satishpasumarthi) [reply README.md]: This is okay with me.

Comment (satishpasumarthi) [reply README.md]: nit: I noticed that you are using `GUnicorn`, but usually people write `Gunicorn` unless they split the word like `Green Unicorn`. Let's move to `Gunicorn` to make it consistent ?

Comment (matherit) [reply README.md]: Good catch. Fixed.

**Review by satishpasumarthi — APPROVED** [review]
LGTM !

**Pull request merged**